tests/integration_tests/functional/test_uffd.py

```diff
@@ -0,0 +1,238 @@
+            'type': SnapshotMemBackendType.UFFD,
+            'path': 'inexsistent'
+        },
+        snapshot_path=jailed_vmstate
+    )
+
+    assert vm.api_session.is_status_bad_request(response.status_code)
+    assert "Load microVM snapshot error: Cannot connect to UDS in order to " \
+           "send information on handling guest memory page-faults due to: " \
+           "No such file or directory (os error 2)" in response.text
+
+
+def test_unbinded_socket(bin_cloner_path, test_microvm_with_api):
+    """
+    Test error scenario when PF handler has not yet called bind on socket.
+
+    @type: negative
+    """
+    logger = logging.getLogger("uffd_unbinded_socket")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm = test_microvm_with_api
+    vm.spawn()
+    jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+
+    socket_path = os.path.join(vm.path, "firecracker-uffd.sock")
+    run_cmd("touch {}".format(socket_path))
+    jailed_sock_path = vm.create_jailed_resource(socket_path)
+
+    response = vm.snapshot.load(
+        mem_backend={
+            'type': SnapshotMemBackendType.UFFD,
+            'path': jailed_sock_path
+        },
+        snapshot_path=jailed_vmstate
+    )
+
+    assert vm.api_session.is_status_bad_request(response.status_code)
+    assert "Load microVM snapshot error: Cannot connect to UDS in order to" \
+           " send information on handling guest memory page-faults due to: " \
+           "Connection refused (os error 111)" in response.text
+
+
+def test_valid_handler(bin_cloner_path,
+                       test_microvm_with_api,
+                       uffd_handler_paths):
+    """
+    Test valid uffd handler scenario.
+
+    @type: functional
+    """
+    logger = logging.getLogger("uffd_unbinded_socket")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm_builder = MicrovmBuilder(bin_cloner_path)
+    vm = test_microvm_with_api
+    vm.spawn()
+
+    # Copy snapshot memory file into chroot of microVM.
+    jailed_mem = vm.create_jailed_resource(snapshot.mem)
+    # Copy the valid page fault binary into chroot of microVM.
+    jailed_handler = vm.create_jailed_resource(
+        uffd_handler_paths['valid_handler']
+    )
+
+    # Spawn page fault handler process.
+    pf_handler_process = spawn_pf_handler(vm, jailed_handler, jailed_mem)
+
+    vm, _ = vm_builder.build_from_snapshot(snapshot, vm=vm,
+                                           resume=True,
+                                           uffd_path=SOCKET_PATH)
+
+    ssh_connection = net_tools.SSHConnection(vm.ssh_config)
+
+    # Verify if guest can run commands.
+    exit_code, _, _ = ssh_connection.execute_command("sync")
+    assert exit_code == 0
+
+    # Kill page fault handler process.
+    pf_handler_process.kill()
```

Comment:
If any previous assert fails, it will not be killed. We could have a PfHandler class that kills the process when going out of scope.